src/pool/srv_pool.c

```diff
@@ -5588,16 +5593,25 @@ ds_pool_elect_dtx_leader(struct ds_pool *pool, daos_unit_oid_t *oid,
 	if (rc != 0)
 		goto out;
 
-	rc = pl_select_leader(oid->id_pub, oid->id_shard / layout->ol_grp_size,
-			      layout->ol_grp_size, tgt_id,
-			      pl_obj_get_shard, layout);
-	pl_obj_layout_free(layout);
-	if (rc < 0)
+	oca = daos_oclass_attr_find(oid->id_pub, NULL);
+	leader_idx = pl_select_leader(oid->id_pub, oid->id_shard / daos_oclass_grp_size(oca),
+				      layout->ol_grp_size, tgt_id, pl_obj_get_shard, layout);
```

Comment:
"oid->id_shard / layout->ol_grp_size" is used to calculate the group index, that is object based, not obj_class based. Because the object layout may be dynamically changed. If we replace "layout->ol_grp_size" with "daos_oclass_grp_size(oca)", we may get the leader from a different group that the update should not be written to.

Reply:
I feel a little confuse as well.
But ""oid->id_shard / layout->ol_grp_size" is used to calculate the group index" @Nasf-Fan do you mean which place uses that?
in cli_obj.c's obj_dkey2shard(), the "idx = hash % grp_size + grp_idx * grp_size" and later in obj_grp_leader_get() it use "idx / obj->cob_grp_size" as grp_idx that case seems fine?

Reply:
In pl_select_leader(), "start = grp_idx * grp_size;". So if "grp_idx" is wrong, then "start" will be wrong, and then the leader maybe from another group.

Reply:
the "oid->id_shard / daos_oclass_grp_size(oca)" looks correct for calculate the grp_idx (the original "oid->id_shard / layout->ol_grp_size" looks incorrect if the grp_size extended and > daos_oclass_grp_size(oca)).

But using daos_oclass_grp_size(oca) seems with a possible problem if "ca_rp_nr == DAOS_OBJ_REPL_MAX", let's see @wangdi1 's comment.

Reply:
For example, the "daos_oclass_grp_size(oca)" is 3, but "layout->ol_grp_size" may be 4. Assume "oid->id_shard" is 24, then with the patch, the "grp_idx" is 8 and "start" for leader is 8 * 4 = 32, that is wrong. 32 may be out of boundary.

Reply:
in that case, as id_shard is 24, then at least should have 9 groups (grp_idx is 0 ~ 8, first id_shard of group 8 is 24).
if "layout->ol_grp_size" extended to 4, just adding one extra shard at the end of each group, but within each group, the valid id_shard within each group is still [grp_idx * 3, (grp_idx + 1) * 3 - 1], but shard_idx is [grp_idx * 4, (grp_idx + 1) * 4 - 1].

So for id_shard 24, its grp_idx is 8, and start shard_idx is 32 but it is valid as long as shard_idx < grp_nr * 4 (36)?

Reply:
id_shard is the real shard of the object, instead of the shard index within the layout, i.e. even VOS will use the id_shard for the object.

so we will use id_shard / daos_oclass_grp_size() to calculate the group index, instead of ol_grp_size, which can be changed during reintegration.

if the grp_size is 4, then 32 is not out of boundary? @Nasf-Fan  what do you mean "start" here?

Note:  pl_select_leader() will return leader shard index, which is not the id_shard, during the reintegration.
So we have to convert it to real shard, then we can compare it with o_id.id_shard.

@liuxuezhao  this is the real object ID, which will never be MAX. The MAX will be turn to domain_nr when creating the object, see daos_oclass_fit_max().

Reply:
Some confused. For the case of object layout is extended from 3 to 4, what means for "id_shard" is 24? It is the No.25 shard or which index of such shard?

Reply:
id_shard is the real shard of the object, which will be stored in OI table, i.e. it will not change during reintegration.   

shard_idx is the offset of shard within the object layout, which might change if the layout is extending.

Reply:
So in this example, the "shard_idx" for "id_shard 24" is 32, right?

Reply:
Then, obj_grp_leader_get() also needs similar changes?